tests/test_env.py

```diff
@@ -232,18 +232,110 @@ def test_process_env_install_from_index(
-def test_process_env_install_from_index_not_found(
-    tmpvenv_active: tuple[py.path.local, py.path.local], local_module_package_index: str
+@pytest.mark.parametrize_any("use_extra_indexes_env", [True, False])
+@pytest.mark.parametrize_any("use_extra_indexes", [True, False])
+@pytest.mark.parametrize_any("use_system_config", [True, False])
+def test_process_env_install_from_index_not_found_env_var(
+    tmpvenv_active: tuple[py.path.local, py.path.local],
+    monkeypatch,
+    create_local_package_index_factory,
+    use_extra_indexes: bool,
+    use_extra_indexes_env: bool,
+    use_system_config: bool,
 ) -> None:
     """
-    Attempt to install a package that does not exist from a pip index. Assert the appropriate error is raised.
+    Attempt to install a package that does not exist from the pip indexes defined in the env vars, in the pip config or in both.
+    This if the system config are used or not.
+    Assert the appropriate error is raised.
     """
-    with pytest.raises(env.PackageNotFound, match="Packages this-package-does-not-exist were not found in the given indexes."):
-        # pass use_system_config=False for security reasons (anyone could publish this package to PyPi)
+    index_url = create_local_package_index_factory()
+
+    extra_index_urls = []
+    if use_extra_indexes:
+        extra_index_1 = create_local_package_index_factory("extra1")
+        extra_index_2 = create_local_package_index_factory("extra2")
+        extra_index_urls = [extra_index_1, extra_index_2]
+
+    extra_index_urls_env = ""
+    if use_extra_indexes_env:
+        extra_index_env1 = create_local_package_index_factory("extra_env1")
+        extra_index_env2 = create_local_package_index_factory("extra_env2")
+        extra_index_urls_env = f"{extra_index_env1} {extra_index_env2}"
+
+    monkeypatch.setenv("PIP_EXTRA_INDEX_URL", extra_index_urls_env)
+
+    expected_indexes = index_url
+    if use_system_config:
+        env_index_url_env = ", " + extra_index_urls_env.replace(" ", ", ") if use_extra_indexes_env else ""
+        # PIP_EXTRA_INDEX_URL expects the different urls to be given as a string with a single space separating them,
+        # while the output has commas separating the different values.
+        expected_indexes = expected_indexes + env_index_url_env
+
+    expected_indexes = expected_indexes + ", " + ", ".join(extra_index_urls) if extra_index_urls else expected_indexes
+
+    expected: str = (
+        "Packages this-package-does-not-exist were not "
+        "found in the given indexes. (Looking in indexes: %s)" % expected_indexes
+    )
+    with pytest.raises(env.PackageNotFound, match=re.escape(expected)):
         env.process_env.install_for_config(
             [Requirement.parse("this-package-does-not-exist")],
+            config=PipConfig(index_url=index_url, extra_index_url=extra_index_urls, use_system_config=use_system_config),
+        )
+
+
+@pytest.mark.slowtest
+def test_process_env_install_with_no_index_url(
```

Comment:
I now realize that this is a dangerous test, because it opens the door for a dependency confusion attack. We should remove it.